### PR TITLE
Costume editor snapping addon

### LINKED ISSUE #3639: paint editor snapping
this would probably take a considerable amount of effort, here are my ideas for features:

- snap to horizontal / vertical center
- align things with other things
  - horizontal centers
  - vertical centers
  - both (put two things in the same place)
  - top, bottom, left, right edges
  - closest edges (making things flush with one another)

might try to do this myself once i figure out scratch and such idk

sorry for probably wasting someone's time

### PR DESCRIPTION
**Pull request by ErrorGamer2000**
Resolves #3639

### Changes

- Addon API Changes
  - Added new shared space to `addon.tab.appendToSharedSpace` to allow addons to add controls in the paint editor.
- Addon Changes
  - Made minor visual changes to the `onion-skinning` addon.
  - Changed the `onion-skinning` settings page positioning. No longer absolute, always centered over button.
  - Changed `onion-skinning` to use the new shared space.
- New Addon
  - Added the new `paint-snap` addon:
    - Allows users to enable/disable snapping altogether, including the native snap to center.
    - Allows users to customize the snapping threshold (or the minimum distance that the selection can move to snap).
    - Allows users to enable/disable snapping to:
      - Edges of the art board / page
      - Corners of the art board / page
      - Center of the art board / page, including the axes
      - Edges of other objects on the page (Groups are considered to be a single object)
      - Centers of other objects on the page
      - Corners of other objects on the page
    - Allows users to enable/disable snapping from:
      - Center of the selection
      - Corners of the selection
      - Centers of the edges of the selection (Really just acts like the full edge, kinda)
    - Works in both vector and bitmap modes.

### Reason for changes

Paint snapping can be extremely useful in other svg editors, so why not add it to scratch.

### Tests

Can't upload video, but tested all options and snapping, as well as changes to the `onion-skinning` UI as editing was being done as well as when it was finished.

### CODE REVIEW

File: addon-api/content-script/Tab.js
Comment (apple502j) [inline addon-api/content-script/Tab.js:359]: Please add the description.

File: addons/paint-snap/addon.json
Comment (apple502j) [inline addons/paint-snap/addon.json:6]:
```suggestion
  "versionAdded": "1.29.0",
```

File: addons/paint-snap/addon.json
Comment (apple502j) [inline addons/paint-snap/addon.json:3]: Needs a better name.

File: addons/paint-snap/addon.json
Comment (apple502j) [inline addons/paint-snap/addon.json:4]:
```suggestion
  "description": "Snap objects in the costume editor to bounding boxes and vector nodes.",
```

File: addons/paint-snap/helper/math.js
Comment (apple502j) [inline addons/paint-snap/helper/math.js:21]:
```suggestion
    return Math.random() >= 0.5;
```

File: addons/paint-snap/lib/keyMirror.js
Comment (apple502j) [inline addons/paint-snap/lib/keyMirror.js:1]: Don't think this is useful. I know this is a copy from vanilla, but don't need this ourselves.

File: addons/paint-snap/state.js
Comment (apple502j) [inline addons/paint-snap/state.js:57]: Please don't.

File: addons/paint-snap/ui.js
Comment (apple502j) [inline addons/paint-snap/ui.js:55]: Instead of this, make `onion-skinning` detect this addon. Do not use `getEnabledAddons` for this. (I'll probably make this crash if this is called from other addons at some point.)

**Review by apple502j — CHANGES_REQUESTED** [review]
Also, please fix the code spaghetti. One big file, and a small file for splittable parts of code. Please don't use module-scoped variable except for addon compatibility.

File: addons/paint-snap/ui.js
Comment (ErrorGamer2000) [reply addons/paint-snap/ui.js:55]: I dont even need this, will remove. Sorry. Also will remove `window.paper = paper`

**Comment by Norbiros** [comment]
It is great addon, but I have couple of suggestions:
1. Why you have options in addon's settings and in own UI. I think that we should remove addon's settings and stay with own ui (you can save selected options in localStorage (something like in `messages filter` addon)).
2. Why you have so much options enabled by default. We should add default configuration for example:
Screenshot from 2022-09-01 20-55-02
3. Maybe you can make icons more grey?
Screenshot from 2022-09-01 21-00-05
4. I can type values like: -5, 1000 or letters to `snapping distance` input.
5. I have console error: `Uncaught TypeError: hideGuides is not a function updateSelectTool.js:545:5`

**Comment by ErrorGamer2000** [comment]
> It is great addon, but I have couple of suggestions:
> 
> 1. Why you have options in addon's settings and in own UI. I think that we should remove addon's settings and stay with own ui (you can save selected options in localStorage (something like in `messages filter` addon)).
> 2. Why you have so much options enabled by default. We should add default configuration for example:
> 3. Maybe you can make icons more grey?
> 4. I can type values like: -5, 1000 or letters to `snapping distance` input.
> 5. I have console error: `Uncaught TypeError: hideGuides is not a function updateSelectTool.js:545:5`
> 6. Can you add dynamic disable/enable.

@Norbiros Thanks for bringing this to my attention.

1. This is a better idea, working on it now.
2. I just set them all to b enabled by default. That actually looks like a good default.
3. I don't know why your zoom controls are like that... 
image
I actually made them the same exact color as the `onion-skinning` icons.
4. Fixed it.
5. Also fixed. Forgot to remove some code.
6. Forgot to add it to the manifest. On it.

**Comment by WorldLanguages** [comment]
I think a toggle would be better than "on/off" because those 2 words, when translated into other languages, can become pretty long.

**Comment by WorldLanguages** [comment]
> Maybe you can make icons more grey?

@Norbiros Are you using the vanilla colors preset of the editor dark mode addon? Looks like ErrorGamer can't reproduce that.

**Comment by ErrorGamer2000** [comment]
> I think a toggle would be better than "on/off" because those 2 words, when translated into other languages, can become pretty long.

I think I'll do a checkbox instead. Probably would be better than a toggle because I don't think that a toggle will match the scratch UI. I was just doing the text toggles because I wanted to match onion skinning.

**Comment by mxmou** [comment]
> I don't know why your zoom controls are like that...
> image
> I actually made them the same exact color as the `onion-skinning` icons.

Onion skinning uses a dark color because the icons have thin strokes and transparency. Zoom controls use `#575e75`.

**Comment by ErrorGamer2000** [comment]
Oh, okay. Change it soon.

**Comment by ErrorGamer2000** [comment]
Fixed

**Review by Norbiros — CHANGES_REQUESTED** [review]
1. When I am dynamic enabling addon I have this error:
```log
Uncaught TypeError: state is undefined [userscript.js:124:7]
```
2. *Not really a bug, but:* When I am changing value using + and - buttons, it selects text input:
Screenshot from 2022-09-06 19-07-19
3. Can we add setting, or can we just change line colour?

**Comment by ErrorGamer2000** [comment]
> 1. When I am dynamic enabling addon I have this error:
> ```log
> Uncaught TypeError: state is undefined [userscript.js:124:7]
> ```
> 2. *Not really a bug, but:* When I am changing value using + and - buttons, it selects text input:
> Screenshot from 2022-09-06 19-07-19
> 3. Can we add setting, or can we just change line colour?
> 

1. @Norbiros that error is not possible since `userscript.js` does not have a line 124, and no other file that I added uses `state` on line 124.
2. It's a "feature"
3. Sure, have that done in a little while.

**Comment by ErrorGamer2000** [comment]
> 1. When I am dynamic enabling addon I have this error:
> ```log
> Uncaught TypeError: state is undefined [userscript.js:124:7]
> ```
> 2. *Not really a bug, but:* When I am changing value using + and - buttons, it selects text input:
> Screenshot from 2022-09-06 19-07-19
> 3. Can we add setting, or can we just change line colour?
> 

1. @Norbiros that error is not possible since `userscript.js` does not have a line 124, and no other file that I added uses `state` on line 124.
2. It's a "feature"
3. Sure, have that done in a little while.

**Comment by Norbiros** [comment]
LGTM

**Comment by ErrorGamer2000** [comment]
Will fix in an hour or so, but that's minor and causes no issues.

**Comment by apple502j** [comment]
@Norbiros just to make sure, did you have default costume editor color enabled when it errored?

**Comment by ErrorGamer2000** [comment]
@apple502j I have it enabled too. And no error.

**Comment by ErrorGamer2000** [comment]
@apple502j and @WorldLanguages please review again, I think I have all of the bugs worked out.

File: addons/paint-snap/state.js
Comment (TheColaber) [inline addons/paint-snap/state.js:50]: This is inconsistent with onion skinning. @WorldLanguages should we add tons of settings like in onion skinning? Or should we have onion skinning do this? Or maybe we add an API to handle storage?

**Review by TheColaber — CHANGES_REQUESTED** [review]

File: addons/paint-snap/addon.json
Comment (ErrorGamer2000) [reply addons/paint-snap/addon.json:2]: Ah, yes.

File: addons/paint-snap/state.js
Comment (ErrorGamer2000) [reply addons/paint-snap/state.js:50]: I vote api.

Comment (Norbiros) [reply addons/paint-snap/state.js:50]: API was created (in #2843) but it wasn't merged...

Comment (ErrorGamer2000) [reply addons/paint-snap/state.js:50]: Ummm

image

Comment (mxmou) [reply addons/paint-snap/state.js:50]: Related: #1950

Comment (ErrorGamer2000) [reply addons/paint-snap/state.js:50]: somewhat, yes

Comment (WorldLanguages) [reply addons/paint-snap/state.js:50]: I don't think it makes sense to delay the release of this addon until we have a good way to handle cases like this.

I'd say we make this addon handle settings just like onion-skinning for the initial release of this addon.

**Comment by ErrorGamer2000** [comment]
I do have a few small tweaks I think I should make, one of which being renaming both settings with "edge centers" to "edge midpoints".

File: addon-api/content-script/Tab.js
Comment (apple502j) [inline addon-api/content-script/Tab.js:359]:
```suggestion
   * paintEditorZoomControls - before the zoom controls in the paint editor
```

File: addons/paint-snap/addon.json
Comment (apple502j) [inline addons/paint-snap/addon.json:5]:
```suggestion
  "versionAdded": "1.29.0",
```

**Review by apple502j — CHANGES_REQUESTED** [review]

**Review by ErrorGamer2000 — COMMENTED** [review]
done

**Comment by Secret-chest** [comment]
The default config should have all corners and centres and no edges IMO.

**Review by apple502j — APPROVED** [review]

**Comment by GarboMuffin** [comment]
Some nits:

 - I think it would make sense to automatically turn on snapping when someone changes the settings, like what onion skinning does
 - I think it would look better if the black border between sections of the settings was made less dark or removed
 - If all of the "Snap from:" or "Snap to:" options are disabled, I think it would make sense to make the snapping icon look disabled
 - When all of "Snap from:" or "Snap to:" are disabled and someone tries to turn on snapping, perhaps it would make sense to turn on a couple things for them

**Comment by ErrorGamer2000** [comment]
> Some nits:
> 
> * I think it would make sense to automatically turn on snapping when someone changes the settings, like what onion skinning does
> * I think it would look better if the black border between sections of the settings was made less dark or removed
> * If all of the "Snap from:" or "Snap to:" options are disabled, I think it would make sense to make the snapping icon look disabled
> * When all of "Snap from:" or "Snap to:" are disabled and someone tries to turn on snapping, perhaps it would make sense to turn on a couple things for them

All fixed @GarboMuffin

**Comment by lisa-wolfgang** [comment]
https://github.com/ScratchAddons/ScratchAddons/issues/5219 was proposed recently. @ErrorGamer2000, would you be interested in implementing this, or would it be best to create a separate PR? Full disclosure, I don't know how much effort this would take.

**Comment by ErrorGamer2000** [comment]
@lisa-wolfgang That is definitely a viable option. Grid mode would have to be a *separate* type of snapping that you can switch to, however, because the two modes would not work very well together. However, I think it would be best to wait until #5226 is either approved or rejected before making the decision or not, because that would end up with a **very** large number of settings for this addon.

**Comment by mxmou** [comment]
> Grid mode would have to be a _separate_ type of snapping that you can switch to, however, because the two modes would not work very well together.

Does this mean that enabling snap to grid would disable snapping to all other objects?

**Comment by ErrorGamer2000** [comment]
@mxmou yes

**Comment by mxmou** [comment]
> @mxmou yes

Is there a technical reason for that? I don't think users would expect this restriction - Inkscape, for example, doesn't have it.

**Comment by ErrorGamer2000** [comment]
> > @mxmou yes
> 
> Is there a technical reason for that? I don't think users would expect this restriction - Inkscape, for example, doesn't have it.

True, but the paint area is so small in scratch, the two would just stumble over each other.

**Comment by lisa-wolfgang** [comment]
I think it should be fine for the majority of cases, especially with large grid sizes, but what I do in Inkscape in that situation is hide the grid, snap, then show the grid again. If I'm remembering correctly, you add a grid snapping toggle in the editor, so that method should work fine with this addon as well.

**Comment by ErrorGamer2000** [comment]
> I think it should be fine for the majority of cases, especially with large grid sizes, but what I do in Inkscape in that situation is hide the grid, snap, then show the grid again. If I'm remembering correctly, you add a grid snapping toggle in the editor, so that method should work fine with this addon as well.

Hmm ok then

**Review by TheColaber — APPROVED** [review]

File: addons/paint-snap/addon.json
Comment (WorldLanguages) [inline addons/paint-snap/addon.json:4]: Name, description and tags LGTM

**Comment by WorldLanguages** [comment]
@ErrorGamer2000 Any blockers? Maybe we could ask GarboMuffin for an explicit review?

**Comment by ErrorGamer2000** [comment]
No blockers and I think Garbo already did @WorldLanguages

**Comment by ErrorGamer2000** [comment]
Ok then @GarboMuffin pls check

**Comment by WorldLanguages** [comment]
This has 2 approvals, so it can be merged (I'll take a quick look first). We should try to find bugs until the translation process ends in a few days.
Next time, #5384 will be useful to avoid PRs like this going stale for 3 months.

**Comment by WorldLanguages** [comment]
We could also decide to add the "beta" tag anytime before actual release. But strings, such as the addon name and description, can't change after today.

**Comment by RedGuy12** [comment]
image

**Comment by CST1229** [comment]
I think the snapping GUI should be visual, something like this:
image

**Comment by WorldLanguages** [comment]
Re: Jazza (screenshot posted by @RedGuy12)

1. Snapping can already be toggled universally for the session through the magnet icon.
2. The find bar going in front is a bug that deserves its own dedicated issue.
3. The menu not hiding when clicking outside is not a new bug, that's how "onion skinning" menu has worked for years. Also needs a dedicated issue.

**Comment by WorldLanguages** [comment]
> I think the snapping GUI should be visual, something like this:

@CST1229 That would be helpful to understand what each specific setting does. But it could get annoying.

**Comment by ErrorGamer2000** [comment]
> I think the snapping GUI should be visual, something like this: image

@CST1229 I honestly considered this. But it would be both difficult to make, and difficult for users to understand.

**Pull request merged**
